### PR TITLE
[TT-2162] Feature respect key expiration

### PR DESCRIPTION
**Pull request by furkansenharputlu**
This PR defines new fields inside API definition and global config named `session_lifetime_respects_key_expiration`. 

When it is: 
`false`: It will not respect expiration date which is bigger than session lifetime and the key will be physically removed.
`true`: It will respect the expiration which is set in the key level and the key will not be physically removed until expiration time finishes.

if global `session_lifetime_respects_key_expiration` is `true` api level will be ignored

### CODE REVIEW

File: user/session.go [hunk 1]
```diff
@@ -225,16 +225,23 @@ func (s *SessionState) KeyHashEmpty() bool {
 	return s.keyHash == ""
 }
 
-func (s *SessionState) Lifetime(fallback int64, forceGlobalSessionLifetime bool, globalSessionLifetime int64) int64 {
+func (s *SessionState) Lifetime(respectSessionLifeTime bool, fallback int64, forceGlobalSessionLifetime bool, globalSessionLifetime int64) int64 {
+	if respectSessionLifeTime {
+		return s.SessionLifetime
+	}
```
Comment (furkansenharputlu) [inline user/session.go:231]: The brain of the fix is here.

File: user/session.go [hunk 2]
```diff
@@ -225,16 +225,23 @@ func (s *SessionState) KeyHashEmpty() bool {
 	return s.keyHash == ""
 }
 
-func (s *SessionState) Lifetime(fallback int64, forceGlobalSessionLifetime bool, globalSessionLifetime int64) int64 {
+func (s *SessionState) Lifetime(respectKeySessionLifeTime bool, fallback int64, forceGlobalSessionLifetime bool, globalSessionLifetime int64) int64 {
+	if respectKeySessionLifeTime {
+		return s.SessionLifetime
+	}
+
```
Comment (furkansenharputlu) [inline user/session.go:232]: The brain of the fix is here.

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4119/merge
Commit: ce0ed0dd7b0b4cc75fcded19b0a93e4885b6bb80 
Triggered by: pull_request (@furkansenharputlu)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2480441074)

**Comment by Tyk-ITS** [comment]
**API tests result: skipped** :no_entry_sign: 
Branch used: refs/pull/4119/merge
Commit: 5271bfa4263a226328a9d7db5d0449de7083af1d 
Triggered by: pull_request (@furkansenharputlu)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2480469635)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4119/merge
Commit: 39b82e40b7b09569402d19c4b1b5b760c2e1f27e 
Triggered by: pull_request (@furkansenharputlu)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2480453397)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4119/merge
Commit:  
Triggered by: pull_request (@furkansenharputlu)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2480421883)

**Review by tbuchaillot — COMMENTED** [review]
@furkansenharputlu what do you think of changing `respect_key_session_lifetime` for `force_key_session_lifetime` to be more cohesive with the current configs?
Also, do you added some documentation for this? I'm afraid it could be a headache of lifetimes when debugging/chasing a problem with so many configs 😅

**Comment by furkansenharputlu** [comment]
@tbuchaillot we had a call with @buger and the previous implementation is changed. Now, it definitely respects :) `respect_key_expiration`. We don't force it if `session_lifetime` is bigger.

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4119/merge
Commit: d3cb9aac32efb4ea1b415030eefda5b8d59d1514 
Triggered by: pull_request (@furkansenharputlu)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2541328271)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4119/merge
Commit: 712d70f0701f54d840f3392803b1884c7b456416 
Triggered by: pull_request (@furkansenharputlu)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2541332609)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4119/merge
Commit: 8b5167f6d0c412d334d14e3063e7ac3fc111ff3b 
Triggered by: pull_request (@furkansenharputlu)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2541338724)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4119/merge
Commit: a018356f64383010bcd6120cba22eb74246d0fca 
Triggered by: pull_request (@furkansenharputlu)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2541396113)

File: user/session.go [hunk 3]
```diff
@@ -225,22 +225,46 @@ func (s *SessionState) KeyHashEmpty() bool {
...
 
 	if fallback > 0 {
-		return fallback
+		return calculateLifetime(respectKeyExpiration, s.Expires, fallback)
 	}
 
 	return 0
 }
 
+func calculateLifetime(respectExpiration bool, expiration, lifetime int64) int64 {
```
Comment (tbuchaillot) [inline user/session.go:244]: @furkansenharputlu please add some func comments

File: user/session.go [hunk 4]
```diff
@@ -225,22 +225,46 @@ func (s *SessionState) KeyHashEmpty() bool {
 	return s.keyHash == ""
 }
 
-func (s *SessionState) Lifetime(fallback int64, forceGlobalSessionLifetime bool, globalSessionLifetime int64) int64 {
+func (s *SessionState) Lifetime(respectKeyExpiration bool, fallback int64, forceGlobalSessionLifetime bool, globalSessionLifetime int64) int64 {
```
Comment (tbuchaillot) [inline user/session.go:228]: @furkansenharputlu can we add some comment to Lifetime func?

**Comment by tbuchaillot** [comment]
> @tbuchaillot we had a call with @buger and the previous implementation is changed. Now, it definitely respects :) `respect_key_expiration`. We don't force it if `session_lifetime` is bigger. If `session_lifetime` is smaller than `expiration`, it will wait expiration time to physically remove the key from redis.

@furkansenharputlu looks good overall! please add some description/comments and make clear in the code/PR/ticket that `force_global_session_lifetime` and `global_session_lifetime` still override this new behavior.

File: apidef/api_definitions.go [hunk 5]
```diff
@@ -593,6 +593,7 @@ type APIDefinition struct {
 	CustomMiddleware           MiddlewareSection      `bson:"custom_middleware" json:"custom_middleware"`
 	CustomMiddlewareBundle     string                 `bson:"custom_middleware_bundle" json:"custom_middleware_bundle"`
 	CacheOptions               CacheOptions           `bson:"cache_options" json:"cache_options"`
+	RespectKeyExpiration       bool                   `bson:"respect_key_expiration" json:"respect_key_expiration"`
```
Comment (buger) [inline apidef/api_definitions.go:596]: I think it require a better name. Like: `remove_keys_after_expiration`

Comment (furkansenharputlu) [reply apidef/api_definitions.go:596]: If session lifetime is bigger, it doesn’t remove after expiration. It waits for expiration to be done

Comment (buger) [reply apidef/api_definitions.go:596]: In this case name it like: `session_lifetime_respect_key_expiration`, or similar.

**Comment by buger** [comment]
Apart from Tomas comments, can you also explain in words how new logic works (and maybe add it as part of function comment or similar), or ensure that it will be added to final docs for this feature.

In order to enable use key expirtation user still need to enable lifetime by setting non zero value to lifetime, either in API or conf level?

**Comment by furkansenharputlu** [comment]
The comment are added and the parameter names are changed to `session_lifetime_respects_key_expiration` both in api level and global config level.

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4119/merge
Commit: 40c5cde2ca9ce341006974edaac56596ba09ee42 
Triggered by: pull_request (@furkansenharputlu)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2696393469)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4119/merge
Commit: 7ab212162339eb2b7bcc7887e02a4009fecad626 
Triggered by: pull_request (@furkansenharputlu)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2696402555)

**Review by buger — APPROVED** [review]

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/heads/master
Commit: 41ec2f95d2ab59274da59cca1f1da6fcc2ab4bb8 Add flags to make session lifetime respects key expiration (#4119)
Triggered by: push (@furkansenharputlu)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2696666196)

**Pull request merged**